toxcore/events/friend_message.c

```diff
@@ -0,0 +1,183 @@
+}
+uint8_t *tox_event_friend_message_get_message(const Tox_Event_Friend_Message *friend_message)
+{
+    assert(friend_message != nullptr);
+    return friend_message->message;
+}
+
+
+/*****************************************************
+ *
+ * :: add/clear/get
+ *
+ *****************************************************/
+
+
+static Tox_Event_Friend_Message *tox_events_add_friend_message(Tox_Events *events)
+{
+    if (events->friend_messages_size == UINT32_MAX) {
+        return nullptr;
+    }
+
+    if (events->friend_messages_size == events->friend_messages_capacity) {
+        const uint32_t new_friend_messages_capacity = events->friend_messages_capacity * 2 + 1;
+        Tox_Event_Friend_Message *new_friend_messages = (Tox_Event_Friend_Message *)realloc(
+                    events->friend_messages, new_friend_messages_capacity * sizeof(Tox_Event_Friend_Message));
+
+        if (new_friend_messages == nullptr) {
+            return nullptr;
+        }
+
+        events->friend_messages = new_friend_messages;
+        events->friend_messages_capacity = new_friend_messages_capacity;
+    }
+
+    Tox_Event_Friend_Message *const friend_message = &events->friend_messages[events->friend_messages_size];
+    tox_event_friend_message_construct(friend_message);
+    ++events->friend_messages_size;
+    return friend_message;
+}
+
+void tox_events_clear_friend_messages(Tox_Events *events)
+{
+    if (events == nullptr) {
+        return;
+    }
+
+    for (uint32_t i = 0; i < events->friend_messages_size; ++i) {
+        tox_event_friend_message_destruct(&events->friend_messages[i]);
+    }
+
+    free(events->friend_messages);
+    events->friend_messages = nullptr;
+    events->friend_messages_size = 0;
+    events->friend_messages_capacity = 0;
+}
+
+uint32_t tox_events_get_friend_messages_size(const Tox_Events *events)
+{
+    return events->friend_messages_size;
+}
+
+const Tox_Event_Friend_Message *tox_events_get_friend_message(const Tox_Events *events, uint32_t index)
+{
+    assert(index < events->friend_messages_size);
+    assert(events->friend_messages != nullptr);
+    return &events->friend_messages[index];
+}
+
+
+/*****************************************************
+ *
+ * :: event handler
+ *
+ *****************************************************/
+
+
+void tox_events_handle_friend_message(Tox *tox, uint32_t friend_number, Tox_Message_Type type, const uint8_t *message,
+                                      size_t length, void *user_data)
+{
+    Tox_Events *events = tox_events_alloc(user_data);
+
+    if (events == nullptr) {
+        return;
+    }
+
+    Tox_Event_Friend_Message *friend_message = tox_events_add_friend_message(events);
+
+    if (friend_message == nullptr) {
+        return;
+    }
+
+    tox_event_friend_message_set_friend_number(friend_message, friend_number);
+    tox_event_friend_message_set_type(friend_message, type);
+    tox_event_friend_message_set_message(friend_message, message, length);
+}
```

Comment:
There are some failures happening here that don't propagate up the stack. If we are not signaling failure, then perhaps at least log it? Also now that I think about it, this whole events module is lacking the logger.

Reply:
Fixed. It's now propagated up.